### DISCUSSION TITLE
Combining repositories

### DISCUSSION THREAD

**Discussion by The-Futurist** [community/community#23811] [New to GitHub]
<p>There is a (and it seems not particularly rare) need sometimes to combine (I’ll avoid using the term “merge” for this) two repositories into a single repository so that we can then delete the original two and do all work on the new combined repository.</p>
<p>This is discussed for example in this <a href="http://choly.ca/post/git-merge-to-monorepo/" rel="nofollow noopener">blogpost</a></p>
<p>You’ll see a question I just asked on that blog and I’m repeating that question here:</p>
<hr>
<p><em>This is an interesting post and I have a question, does your solution cover the most genera case? Here’s what I mean.</em></p>
<p><em>I have <strong>repo_a</strong> and <strong>repo_b</strong>.</em></p>
<p>_ <strong>repo_a</strong> has branches master, gui_project and api_project branches_<br>
_ <strong>repo_b</strong> has branches master, mem_project and api_project branches._</p>
<p><em>In this scenario api_project has the same name in each repo because it is the same project, some code in <strong>repo_a</strong> and some in <strong>repo_b</strong>.</em></p>
<p><em>I want to combine these and end up with <strong>repo_c</strong> that has 4 branches:</em></p>
<p><em>master, gui_project, mem_project and api_project.</em></p>
<p><em>where:</em></p>
<p>_master contains all commits from master in <strong>repo_a</strong> and <strong>repo_b</strong> _<br>
_api_project contains all commits from api_project in <strong>repo_a</strong> and <strong>repo_b</strong> _<br>
_mem_project contains all commits from <strong>repo_b</strong> _<br>
_gui_project contains all commits from <strong>repo_a</strong> _</p>
<p><em>The repos are completely different, their folder structure and so on are unrelated so no folder/path/file hieracrchy appears in more than one repo, that is there is no code/files that is in each repo.</em></p>
<p><em>This is my challenge, in repos where we have a lot of branches some unique to each repo and some with the same name in each repo because the work on them pertains to the same project.</em></p>
<hr>
<p>Now I know that GitHub doesn’t provide such a capability (be great if it did as this need often arises from time to time as projects or organizations grow and mature) but has anyone done this, in the general case that I outline above?</p>
<p>In reality there a quite a lot more branches than just four, but this should be academic if we have a general solution.</p>
<p>I want to attempt this but also want to gather as much input from others and git gurus before I actually begin, I’d hate to spend an hour or two then hit some roadblock simply because of a lack of due dilligence.</p>
<p>Thanks</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>:rotating_light:  <strong>WARNING:</strong> At no point in what I describe below did I need to use <code>git push --force</code>. If you find yourself with the urge to use <code>git push --force</code>, then you are at risk of losing data and you should back out and start over. <code>git push --force</code> is a potentially destructive command, use it <strong>at your own risk</strong>. :rotating_light:</p>
<p>Let me start by saying that you should protect yourself from data loss by <a href="https://en.wikipedia.org/wiki/Scratch_monkey" rel="nofollow noopener">using scratch repositories</a> to perform your experiments. A scratch repository is one that is disconnected from all other repos so that you can’t possibly mess up anything but the scratch repo, but the various data sources that you’re using to build the scratch one are guaranteed intact. One of the nice things about git though is that it makes this pretty easy. If at any time you mess something up, just delete the entire local repository and start over.</p>
<p>So here’s what I did, you can see the results in <a href="https://github.com/lee-dohm/merged-repo" rel="nofollow noopener">https://github.com/lee-dohm/merged-repo</a>, which is a combination of <a href="https://github.com/lee-dohm/test-repo" rel="nofollow noopener">https://github.com/lee-dohm/test-repo</a> and <a href="https://github.com/lee-dohm/test-repo-2" rel="nofollow noopener">https://github.com/lee-dohm/test-repo-2</a>. I took your requirements:</p>
<ul>
<li>
<em>n</em> repositories</li>
<li>
<strong>No</strong> duplicate paths between the repositories on any combination of branches</li>
<li>At least one branch name common between <em>m</em> repositories</li>
<li>At least one branch that is unique among all repositories</li>
</ul>
<p>To satisfy these requirements, I crated the following scenario:</p>
<ul>
<li>From lee-dohm/test-repo, I would take branches <code>master</code> and <code>lee-dohm-patch-4</code>
</li>
<li>From lee-dohm/test-repo-2, I would take branches <code>master</code> and <code>branch-that-does-not-exist-on-test-repo</code>
</li>
<li>The <code>master</code> branches from each would need to be merged</li>
<li>The unique branches would each need to contain all contents from the merged <code>master</code> and the unique contents from their source repo</li>
</ul>
<p>First, I created a scratch repo:</p>
<pre><code>mkdir merged-repo
cd merged-repo
git init
</code></pre>
<p>Then, I hooked up the remotes:</p>
<pre><code>git remote add test-repo https://github.com/lee-dohm/test-repo.git
git remote add other-repo https://github.com/lee-dohm/test-repo-2.git
</code></pre>
<p>Then, I needed to bring in all the source refs from the two repositories:</p>
<pre><code>git fetch test-repo master
git fetch other-repo master
git fetch test-repo lee-dohm-patch-4
git fetch other-repo branch-that-does-not-exist-in-test-repo
</code></pre>
<p>At this point, you can protect yourself from altering the source repos by deleting the remotes created:</p>
<pre><code>git remote remove test-repo
git remote remove other-repo
</code></pre>
<p>Here’s how I merged the history of the <code>master</code> branches:</p>
<pre><code>git merge test-repo/master
git merge --allow-unrelated-histories other-repo/master
</code></pre>
<p>The key here, of course, being the <a href="https://git-scm.com/docs/git-merge#Documentation/git-merge.txt---allow-unrelated-histories" rel="nofollow noopener"><code>--allow-unrelated-histories</code> parameter to <code>git merge</code></a>. Now that the two histories have been joined, we can work on the other branches (whether their names are shared or unique).</p>
<pre><code>git checkout -b lee-dohm-patch-4 test-repo/lee-dohm-patch-4
git rebase master
git checkout -b branch-that-does-not-exist-in-test-repo other-repo/branch-that-does-not-exist-in-test-repo
git rebase master
</code></pre>
<p>Rebasing each of the unique branches off of the local <code>master</code> means that they are now based from the shared history from both repositories.</p>
<p>Then I created a new GitHub repository to host the merged results and pushed everything up:</p>
<pre><code>git remote add origin https://github.com/lee-dohm/merged-repo.git
git checkout master
git push origin master
git checkout lee-dohm-patch-4
git push origin lee-dohm-patch-4
git checkout branch-that-does-not-exist-in-test-repo
git push origin branch-that-does-not-exist-in-test-repo
</code></pre>
<p>I hope that helps!</p>

**Comment by The-Futurist** [comment]
<p>Lee, I greatly appreciate this, I will do a similar experiment and follow your directions and if that works out I’ll then try it with the two real repos.</p>
<p>Many thanks.</p>

**Comment by lee-dohm** [comment]
<p>3 posts were split to a new topic: <a href="https://github.community/t/merge-unrelated-histories-and-keep-parts-unrelated/116404">Merge unrelated histories and keep parts unrelated</a></p>

**Comment by Faramarzi123** [comment]
آیا شما خارش دارید؟
